integrations/Matomo/util.js

```diff
@@ -0,0 +1,332 @@
+import get from "get-value";
+import each from "@ndhoule/each";
+import logger from "../../utils/logUtil";
+import { getHashFromArray, getIntegrationsObj } from "../utils/commonUtils";
+
+/** If any event name matches with the goals list provided by the dashboard
+ * we will call the trackGoal with the id provided in the mapping.
+ * @param  {} event
+ * @param  {} goalListMap
+ */
+const goalIdMapping = (event, goalListMap, message) => {
+  const revenue = get(message, "properties.revenue");
+  each((val, key) => {
+    if (key === event.toLowerCase()) {
+      val.forEach((v) => {
+        if (revenue) window._paq.push(["trackGoal", v, revenue]);
+        else window._paq.push(["trackGoal", v]);
+      });
+      // window._paq.push(["trackGoal", val]);
+    }
+  }, goalListMap);
+};
+
+/** Mapping Standard Events 
+ If any event name matches with the standard events list provided in the dashboard
+ @param  {} event
+ @param  {} standardEventsMap
+ @param  {} message
+ */
+const standardEventsMapping = (event, standardEventsMap, message) => {
+  each((val, key) => {
+    if (key === event.toLowerCase()) {
+      let url;
+      let linkType;
+      let keyword;
+      let category;
+      let resultsCount;
+      let contentInteraction;
+      let contentName;
+      let contentPiece;
+      let contentTarget;
+      let domId;
+      const properties = get(message, "properties");
+      val.forEach((v) => {
+        switch (v) {
+          case "trackLink":
+            url = get(properties, "url") || get(message, "context.page.url");
```

Comment:
We can reduce the use of `get` by first checking if `properties` is valid object from message. Then we can use . notation to extract those, for js-sdk `get` is very expensive function whcih will cause performance issue.

So if `properties` is undefined and we try to do `properties.url` we will get an excepion but if we have `properties` defined and we it doesn't have `url` , `url` = `properties.url` , `url` will be `undefined` which is fine. Lets not use `get` here

Reply:
Done